ex2_system/source/logger/logger.c

```diff
@@ -13,145 +13,218 @@
+        }
+    }
 
-bool circular_buf_empty(cbuf_handle_t cbuf)
-{
-    configASSERT(cbuf);
+    if (input_queue == NULL) {
+        do_output(buffer);
+        return;
+    } else {
+        xQueueSend(input_queue, buffer, 0); // will not wait for queue to have space
+    }
 
-    return (!cbuf->full && (cbuf->head == cbuf->tail));
+    return;
 }
 
-size_t circular_buf_capacity(cbuf_handle_t cbuf)
-{
-    configASSERT(cbuf);
-
-    return cbuf->max;
-}
+/**
+ * @brief
+ * Puts a string on the output
+ * 
+ * @details
+ * If filesystem is initialized it will write to the syslog.
+ * If the MCU it is being run on is a flatsat it will not print on the uart
+ * Prepends the uptime in seconds
+ * 
+ * @param String to print
+ * @return None
+ */
+static void do_output(const char *str) {
+    char output_string[STRING_MAX_LEN] = {0};
 
-size_t circular_buf_size(cbuf_handle_t cbuf)
-{
-    configASSERT(cbuf);
+    uint32_t uptime = (uint32_t)(xTaskGetTickCount()/configTICK_RATE_HZ);
 
-    size_t size = cbuf->max;
+    snprintf(output_string, STRING_MAX_LEN, "[%010d]%s\r\n", uptime, str);
 
-    if(!cbuf->full)
-    {
-        if(cbuf->head >= cbuf->tail)
-        {
-            size = (cbuf->head - cbuf->tail);
-        }
-        else
-        {
-            size = (cbuf->max + cbuf->head - cbuf->tail);
-        }
+    if (fs_init) {
+        red_write(logger_file_handle, output_string, strlen(output_string));
     }
 
-    return size;
+#ifndef IS_FLATSAT
+    printf(output_string);
+#endif
 }
 
-static void advance_pointer(cbuf_handle_t cbuf)
-{
-    configASSERT(cbuf);
-
-    if(cbuf->full)
-    {
-        cbuf->tail = (cbuf->tail + 1) % cbuf->max;
+/**
+ * @brief
+ * Initialize filesystem related contructs for logger
+ * 
+ * @details
+ * Opens syslog file and sets fs_init to true on success
+ */
+bool init_logger_fs() {
+    if (fs_init) {
+        return true;
     }
 
-    cbuf->head = (cbuf->head + 1) % cbuf->max;
-    cbuf->full = (cbuf->head == cbuf->tail);
-}
-
-static void retreat_pointer(cbuf_handle_t cbuf)
-{
-    configASSERT(cbuf);
-
-    cbuf->full = FALSE;
-    cbuf->tail = (cbuf->tail + 1) % cbuf->max;
+    int32_t fd = red_open(logger_file, RED_O_RDWR | RED_O_APPEND);
+    if (fd == -1) {
+        int errno = red_errno;
+        fd = red_open(logger_file, RED_O_CREAT | RED_O_RDWR);
+        if (fd == -1) {
+            errno = red_errno;
+            return false;
+        }
+    }
+    fs_init = true;
+    logger_file_handle = fd;
+    return true;
 }
 
-void circular_buf_put(cbuf_handle_t cbuf, uint8_t data)
-{
-    configASSERT(cbuf && cbuf->buffer);
-
-    cbuf->buffer[cbuf->head] = data;
-
-    advance_pointer(cbuf);
+/**
+ * @brief
+ * Shut down filesystem related constructs for logger
+ * 
+ * @details
+ * Closes syslog file and sets fs_init to false
+ */
+void stop_logger_fs() {
+    if (!fs_init) {
+        return;
+    }
+    red_close(logger_file_handle);
+    logger_file_handle = 0;

```

Comment:
Investigate this - do we need to protect this?

Reply:
I don't think so